src/iter/test.rs

```diff
@@ -1608,3 +1608,14 @@ fn check_partition_map() {
     assert_eq!(a, vec![1, 2, 3]);
     assert_eq!(b, "abcxyz");
 }
+
+#[test]
+fn check_repeat_eq() {
+    let v = vec!(4,4,4,4);
+    let mut fours: Vec<_> = repeat(4)
+    .take(4)
+    .zip(v)
+    .collect();
```

Comment:
The producer will need to track its effective length.  It should initialize to `usize::MAX` since that's what `len()` reported.  Then each time it splits, use the given index to determine the resulting length of each side.  And finally the iterator will also need to limit its length.